### PR TITLE
Refactor the rsr-reader and add more test coverage

### LINKED ISSUE #122: Pyspectral incompatible with h5py=3.1.0
#### Code Sample, a minimal, complete, and verifiable piece of code

```python
# Your code here
>>> from pyspectral.rsr_reader import RelativeSpectralResponse
>>> rsr = RelativeSpectralResponse('NOAA-20', 'viirs')
```
#### Problem description
The code fails for latest versions of h5py (3.1.0 released Nov 6, 2020), see below!

#### Expected Output
Code should run without exceptions and return the RSR data for N20/VIIRS

#### Actual Result, Traceback if applicable
```
rsr data may not be up to date: /data/proj/safutv/.local/share/pyspectral/rsr_viirs_NOAA-20.h5
4494378it [00:02, 1923809.28it/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/san1/opt/viirs_imagery_processor/releases/viirs_imagery_processor-0.2.65/lib/python3.8/site-packages/pyspectral/rsr_reader.py", line 125, in __init__
    self.load()
  File "/san1/opt/viirs_imagery_processor/releases/viirs_imagery_processor-0.2.65/lib/python3.8/site-packages/pyspectral/rsr_reader.py", line 174, in load
    self.description = self.description.decode('utf-8')
AttributeError: 'str' object has no attribute 'decode'
```

#### Versions of Python, package at hand and relevant dependencies
`h5py=3.1.0`
`pyspectral=0.10.1` and latest master as of Nov 10, 2020

Thank you for reporting an issue !

### PR DESCRIPTION
**Pull request by adybbroe**
Signed-off-by: Adam.Dybbroe <a000680@c21856.ad.smhi.se>

 - [x] Closes #122  
 - [x] Tests added 
 - [x] Tests passed: Passes ``pytest pyspectral`` 
 - [x] Passes ``flake8 pyspectral`` 
 - [ ] Fully documented

### CODE REVIEW

File: pyspectral/tests/test_rsr_reader.py [hunk 1]
```diff
@@ -167,3 +167,68 @@ def test_integral(self, get_rsr_version, download_rsr, load, isfile, exists):
...
+    def test_set_platform_name(self, check_instrument, get_filename, check_filename_exist, load):
+        """Test setting the platform name."""
+
+        load.return_value = None
+        check_filename_exist.return_value = None
+        get_filename.return_value = None
+        check_instrument.return_value = None
+
+        #test_rsr = RelativeSpectralResponse('EOS-Aqua', 'modis')
+        #modis_aqua_filepath = "/home/a000680/data/pyspectral/rsr_modis_EOS-Aqua.h5"
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/tests/test_rsr_reader.py:253]: These comments should be removed

File: pyspectral/tests/test_rsr_reader.py [hunk 2]
```diff
@@ -167,3 +167,68 @@ def test_integral(self, get_rsr_version, download_rsr, load, isfile, exists):
...
+    @patch('pyspectral.rsr_reader.RelativeSpectralResponse._check_instrument')
+    def test_set_platform_name(self, check_instrument, get_filename, check_filename_exist, load):
+        """Test setting the platform name."""
+
+        load.return_value = None
+        check_filename_exist.return_value = None
+        get_filename.return_value = None
+        check_instrument.return_value = None
+
+        #test_rsr = RelativeSpectralResponse('EOS-Aqua', 'modis')
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/tests/test_rsr_reader.py]: These comment should be removed

[the remark this answers is not in the document]
Comment (adybbroe) [reply pyspectral/tests/test_rsr_reader.py]: Done!

File: pyspectral/tests/test_rsr_reader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adybbroe) [reply pyspectral/tests/test_rsr_reader.py]: Done - you are too quick for me @mraspaud ! :-)
I saw them after I pushed.

File: pyspectral/rsr_reader.py [hunk 3]
```diff
@@ -262,6 +216,87 @@ def convert(self):
...
+
+                try:
+                    central_wvl = h5f[bandname][
+                        dname].attrs['central_wavelength']
+                except KeyError:
+                    central_wvl = h5f[bandname].attrs['central_wavelength']
+
+                self.rsr[bandname][dname][
+                    'central_wavelength'] = central_wvl
+
```
Comment (mraspaud) [inline pyspectral/rsr_reader.py:299]: This function should be refactored, it is too long and cryptic

**Review by mraspaud — COMMENTED** [review]
LGTM, I just would like to see one function refactored :)

File: pyspectral/rsr_reader.py (hunk 3 above)
Comment (adybbroe) [reply pyspectral/rsr_reader.py:299]: I agree! I will do!

File: pyspectral/rsr_reader.py [hunk 4]
```diff
@@ -262,6 +212,105 @@ def convert(self):
...
+        self.description = np2str(h5f.attrs['description'])
+        #self.description = h5f.attrs['description']
+        # if isinstance(self.description, bytes):
+        #    self.description = self.description.decode('utf-8')
+
+    def set_band_names(self, h5f):
+        """Set the band names."""
+        self.band_names = h5f.attrs['band_names']
+        self.band_names = [np2str(x) for x in self.band_names]
+        #self.band_names = h5f.attrs['band_names'].tolist()
```
[the remark this answers is not in the document]
Comment (adybbroe) [reply pyspectral/rsr_reader.py]: Fixed

File: pyspectral/rsr_reader.py [hunk 5]
```diff
@@ -262,6 +212,105 @@ def convert(self):
             errmsg = "Conversion from {wn} to {wl} not supported yet".format(wn=WAVE_NUMBER, wl=WAVE_LENGTH)
             raise NotImplementedError(errmsg)
 
+    def set_description(self, h5f):
+        """Set the description."""
+
+        self.description = np2str(h5f.attrs['description'])
+        #self.description = h5f.attrs['description']
```
[the remark this answers is not in the document]
Comment (adybbroe) [reply pyspectral/rsr_reader.py]: Done

File: pyspectral/tests/test_rsr_reader.py [hunk 6]
```diff
@@ -167,3 +167,68 @@ def test_integral(self, get_rsr_version, download_rsr, load, isfile, exists):
...
+        viirs_noaa20_filepath = "/mypath/myfilename"
+        test_rsr = RelativeSpectralResponse(filename=viirs_noaa20_filepath)
+        test_rsr.filename = viirs_noaa20_filepath
+
+        test_rsr.set_platform_name(self.h5f_noaa20_viirs)
+        self.assertEqual(test_rsr.platform_name, 'NOAA-20')
+
+        import ipdb
+        ipdb.set_trace()
+        x = 1
```
[the remark this answers is not in the document]
Comment (adybbroe) [reply pyspectral/tests/test_rsr_reader.py]: Done!

**Comment by adybbroe** [comment]
Ok to merge @pnuu ?
I don't see how I should improve the deepcode issue. And I think the stickler issues should be gone...

File: pyspectral/tests/test_rsr_reader.py [hunk 7]
```diff
@@ -167,3 +168,150 @@ def test_integral(self, get_rsr_version, download_rsr, load, isfile, exists):
...
+    def test_create_rsr_instance(self, check_instrument, get_filename, check_filename_exist, load):
+        """Test setting the platform name."""
+
+        load.return_value = None
+        check_filename_exist.return_value = None
+        get_filename.return_value = None
+        check_instrument.return_value = None
+
+        with pytest.raises(AttributeError) as exec_info:
+            dummy = RelativeSpectralResponse('MyPlatform')
```
[the remark this answers is not in the document]
Comment (pnuu) [reply pyspectral/tests/test_rsr_reader.py:247]: Replacing `dummy` with `_` will make Stickler happy.

[the remark this answers is not in the document]
Comment (djhoese) [reply pyspectral/tests/test_rsr_reader.py:247]: This is still an issue. If you don't use it, assign to `_` or don't assign it at all.

File: pyspectral/tests/test_utils.py [hunk 8]
```diff
@@ -23,11 +23,14 @@
...
 """Do the unit testing for the utils library."""
 
-from pyspectral import utils
-
+import pytest
 import unittest
 import numpy as np
 
+from pyspectral import utils
+from pyspectral.utils import np2str, bytes2string
```
Comment (pnuu) [inline pyspectral/tests/test_utils.py:31]: I think the functions that are tested should be imported inside the test function/method. That way they can be mocked in other tests, if it becomes necessary.

File: pyspectral/tests/test_rsr_reader.py [hunk 9]
```diff
@@ -167,3 +171,279 @@ def test_integral(self, get_rsr_version, download_rsr, load, isfile, exists):
...
+
+        h5f = {}
+        for name in self.viirs_bandnames:
+            h5f[name] = {}
+            for det_name in self.viirs_detector_names:
+                h5f[name][det_name] = self.viirs_rsr_data
+
+        test_rsr.rsr['M1'] = {'det-1': {}}
+        test_rsr.set_band_central_wavelength_per_detector(h5f, 'M1', 'det-1')
+        self.assertTrue('central_wavelength' in test_rsr.rsr['M1']['det-1'])
```
Comment (djhoese) [inline pyspectral/tests/test_rsr_reader.py:391]: You could use `assertIn` here to make deepcode happier: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertIn

Right?

**Review by pnuu — APPROVED** [review]
Just couple of comments. Feel free to merge.

File: pyspectral/tests/test_rsr_reader.py (hunk 9 above)
Comment (adybbroe) [reply pyspectral/tests/test_rsr_reader.py]: Oh, yes, right! thx!

File: pyspectral/tests/test_utils.py (hunk 8 above)
Comment (adybbroe) [reply pyspectral/tests/test_utils.py:31]: Okay, I thought the mocking was independent of the imports actually!?
I always have all imports at the top.

**Pull request merged**